core-bundle/src/Resources/contao/widgets/SerpPreview.php

```diff
@@ -72,115 +90,116 @@ public function generate()
+		return $model->{$this->aliasField};
 	}
 
+	/**
+	 * @todo Use the router to generate the URL in a future version (see #831)
+	 */
 	private function getUrl(Model $model)
 	{
-		if (isset($this->serpPreview['url']))
+		if (!isset($this->url_callback))
 		{
-			return $this->serpPreview['url'];
+			throw new \LogicException('No url_callback given');
 		}
 
-		// FIXME: use the router to generate the URL (see #831)
-		switch (true)
-		{
-			case $model instanceof PageModel:
-				return $model->getAbsoluteUrl();
-
-			case $model instanceof NewsModel:
-				return News::generateNewsUrl($model, false, true);
+		$placeholder = bin2hex(random_bytes(10));
 
-			case $model instanceof CalendarEventsModel:
-				return Events::generateEventUrl($model, true);
+		// Pass a detached clone with the alias set to the placeholder
+		$tempModel = clone $model;
+		$tempModel->origAlias = $tempModel->alias;
+		$tempModel->alias = $placeholder;
+		$tempModel->preventSaving(false);
 
-			default:
-				throw new \RuntimeException(sprintf('Unsupported model class "%s"', \get_class($model)));
+		if (\is_array($this->url_callback))
+		{
+			$this->import($this->url_callback[0]);
+			$url = $this->{$this->url_callback[0]}->{$this->url_callback[1]}($tempModel);
 		}
-	}
-
-	private function getTitleField()
-	{
-		if (!isset($this->serpPreview['title']))
+		elseif (\is_callable($this->url_callback))
 		{
-			return 'ctrl_title';
+			$url = \call_user_func($this->url_callback, $tempModel);
 		}
-
-		if (\is_array($this->serpPreview['title']))
+		else
 		{
-			return 'ctrl_' . $this->serpPreview['title'][0];
+			throw new \LogicException('Please provide the url_callback as callable');
 		}
 
-		return 'ctrl_' . $this->serpPreview['title'];
+		return str_replace($placeholder, '%s', $url);
 	}
 
-	private function getTitleFallbackField()
+	private function getTitleField($suffix)
 	{
-		if (!isset($this->serpPreview['title']) || !\is_array($this->serpPreview['title']))
+		if (!isset($this->titleFields))
 		{
-			return '';
+			return 'ctrl_title' . $suffix;
 		}
 
-		return 'ctrl_' . $this->serpPreview['title'][1];
+		return 'ctrl_' . $this->titleFields[0] . $suffix;
 	}
 
-	private function getAliasField()
+	private function getTitleFallbackField($suffix)
 	{
-		if (!isset($this->serpPreview['alias']))
+		if (!isset($this->titleFields))
 		{
-			return 'ctrl_alias';
+			return '';
 		}
 
-		return 'ctrl_' . $this->serpPreview['alias'];
+		return 'ctrl_' . $this->titleFields[1] . $suffix;
 	}
 
-	private function getDescriptionField()
+	private function getDescriptionField($suffix)
 	{
-		if (!isset($this->serpPreview['description']))
+		if (!isset($this->descriptionFields))
 		{
-			return 'ctrl_description';
+			return 'ctrl_description' . $suffix;
 		}
 
-		if (\is_array($this->serpPreview['description']))
+		return 'ctrl_' . $this->descriptionFields[0] . $suffix;
+	}
+
+	private function getDescriptionFallbackField($suffix)
+	{
+		if (!isset($this->descriptionFields))
```

Comment:
Should probably check for `$this->descriptionFields[1]`

Reply:
Fixed in 07cc4886e37d05be4b7646f27faa3859e448c98c.